### ISSUE TITLE
Console errors threat

### ISSUE THREAD

**Issue by FrutyX** [DinoDevs/GladiatusCrazyAddon#311]
This error has been there for ages (I think I have seen it in 4.3.4 too), and I've been ignoring it, because it's harmless, but maybe you wish to investigate (polish it).

It's occurs randomly, and can occur on any site (any mod), it's not specific. Since I have enabled developer mode in my browser, I see a lot of these in the stored error logs.

```
gca.tools.js?4.3.5&built=1640588198107:1776 Uncaught TypeError: Cannot read properties of null (reading 'getElementsByClassName')
    at Object.prepareFooterButtons (gca.tools.js?4.3.5&built=1640588198107:1776)
    at Object.footerButton (gca.tools.js?4.3.5&built=1640588198107:1764)
    at Object.settingsLink (gca.tools.js?4.3.5&built=1640588198107:1759)
    at Object.inject (packages.js?4.3.5&built=1640588198110:63)
    at fireLoad (packages.js?4.3.5&built=1640588198110:1089)
    at packages.js?4.3.5&built=1640588198110:1093
    at packages.js?4.3.5&built=1640588198110:1098
```

(`let wrapper = footer.getElementsByClassName('gca-footer-buttons-wrapper');`)

**Comment by FrutyX** [comment]
I would like to have an explanation from Gram (or Apo, doesn't matter) about these.

I guess they can be just some random read errors, harmless but still occurring, should I even report these random errors? Is this any useful to you?

It's hard to reproduce it, as it is truly random. I've tried to turn off uBlock (because I got the idea what if he is interfering) and it took me like 50 visited pages to get this error, so uBlock is not the problem here.

**Comment by FrutyX** [comment]
There is another one (more common), it looks it ONLY appears when fighting an event enemy (currently Icy Turna).

And it seems it's breaking other features, the ring should be colored.

```
reports.js?4.3.5&built=1640595022472:200 Uncaught TypeError: Cannot read properties of null (reading '1')
    at Object.report_found_items (reports.js?4.3.5&built=1640595022472:200)
    at Object.inject (reports.js?4.3.5&built=1640595022472:31)
    at fireLoad (reports.js?4.3.5&built=1640595022472:1104)
```

**Comment by FrutyX** [comment]
@GreatApo This popped up (after the fix), but only one time, so I don't know.

```
Uncaught DOMException: Failed to execute 'appendChild' on 'Node': Only one element on document allowed.
    at Object.prepareFooterButtons (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/gca.tools.js?4.3.5&built=1640610018520:1785:11)
    at Object.footerButton (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/gca.tools.js?4.3.5&built=1640610018520:1764:23)
    at Object.settingsLink (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/gca.tools.js?4.3.5&built=1640610018520:1759:20)
    at Object.inject (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:80:20)
    at fireLoad (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:1843:16)
    at chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:1846:3
    at chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:1851:3
```

**Comment by GreatApo** [comment]
> @GreatApo This popped up (after the fix), but only one time, so I don't know.
> 
> ```
> Uncaught DOMException: Failed to execute 'appendChild' on 'Node': Only one element on document allowed.
>     at Object.prepareFooterButtons (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/gca.tools.js?4.3.5&built=1640610018520:1785:11)
>     at Object.footerButton (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/gca.tools.js?4.3.5&built=1640610018520:1764:23)
>     at Object.settingsLink (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/gca.tools.js?4.3.5&built=1640610018520:1759:20)
>     at Object.inject (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:80:20)
>     at fireLoad (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:1843:16)
>     at chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:1846:3
>     at chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:1851:3
> ```

How can I replicate this? The footer should always be available, this shouldn't be a problem in the first place...

**Comment by FrutyX** [comment]
You can't, it showed only once and I don't see it anymore. Weird. I'll let you know if I see it again.

Maybe it popped up only when I replaced the old files with the fix (browser wasn't expecting the change, cache? And throwed an error? Just guessing)

**Comment by FrutyX** [comment]
Yeah, we are fine, hopefully it's not a 1 out of 50 pages error type. Onto the next one! That one should be easy to reproduce, thus easier to fix.

**Comment by FrutyX** [comment]
> **EDIT: This is present every time when an event enemy drops an item! Confirmed.**

> ```
> reports.js?4.3.5&built=1640595022472:200 Uncaught TypeError: Cannot read properties of null (reading '1')
>     at Object.report_found_items (reports.js?4.3.5&built=1640595022472:200)
>     at Object.inject (reports.js?4.3.5&built=1640595022472:31)
>     at fireLoad (reports.js?4.3.5&built=1640595022472:1104)
> ```

Maybe I should mention that we are time limited on this one, since Icy Tundra will end soon, so if you wish to address it (and test it), do it asap.

I should've mentioned this earlier, sorry.

**Comment by GramThanos** [comment]
> @GreatApo This popped up (after the fix), but only one time, so I don't know.
> 
> ```
> Uncaught DOMException: Failed to execute 'appendChild' on 'Node': Only one element on document allowed.
>     at Object.prepareFooterButtons (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/gca.tools.js?4.3.5&built=1640610018520:1785:11)
>     at Object.footerButton (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/gca.tools.js?4.3.5&built=1640610018520:1764:23)
>     at Object.settingsLink (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/gca.tools.js?4.3.5&built=1640610018520:1759:20)
>     at Object.inject (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:80:20)
>     at fireLoad (chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:1843:16)
>     at chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:1846:3
>     at chrome-extension://hkbmmenmpniffngmkigofhgbflhhokka/core/source/overview.js?4.3.5&built=1640610018522:1851:3
> ```

The only place I think this might occur is during the sever maintenance page.

**Comment by GramThanos** [comment]
> 

If you can please post here the output of
`document.getElementById('defenderAvatar11').getElementsByTagName('div')[2].style.backgroundImage`

**Comment by FrutyX** [comment]
> The only place I think this might occur is during the sever maintenance page.

Impossible, I was on the overview page, I didn't saw the maintenance page.

But it doesn't matter, so forget it, the most important thing is, that it did not occurred again, and the fix is working.

**Comment by GramThanos** [comment]
5ed4c733847f1c80d1c769ef511c409c441dbf04 needs testing

**Comment by FrutyX** [comment]
@GramThanos beep, beep, beep, beeeeeeeeeeeeeeep,

![error](https://user-images.githubusercontent.com/23278552/147662382-ced9f1e5-7ed4-4ecc-92dc-dc1b685b4bd4.PNG)

**Comment by FrutyX** [comment]
![defender](https://user-images.githubusercontent.com/23278552/147662570-eb4407b6-3044-407a-b215-ee9c7b5e3f7a.PNG)

**Comment by FrutyX** [comment]
Wait, i think this is for #303 , nah, whatever :D.

**Comment by FrutyX** [comment]
Okay, so far it seems there is no errors, the event location is good too.

1658d23 did the trick, I think.

**Comment by FrutyX** [comment]
Aaaaaaaand I found another one, it's in combat reports, seems the script has some problem (symbols I guess) with guy's name, it's `*matesov*`

**`Uncaught SyntaxError: Invalid regular expression: /*matesov*/: Nothing to repeat`**

`https://s34-cz.gladiatus.gameforge.com/game/index.php?mod=reports&submod=showCombatReport&t=3&reportId=4169146&sh=`

`core/source/reports.js?4.3.5&built=1640944080423:821`

it's

`if ( text.match(playerI) && n > text.indexOf(playerI) ){`

**Comment by FrutyX** [comment]
@GreatApo you ready for another one? Haha.

`Uncaught TypeError: Cannot read properties of undefined (reading 'dataset')`
`https://s34-cz.gladiatus.gameforge.com/game/index.php?mod=guildStorage&doll=6&subsub=0&sh=`

`core/source/guild.storage.js?4.3.5&built=1640958778662:180`
it's
`if (item) delete item[0].dataset.gcaFlag_doubleClickEvent;`

**Comment by FrutyX** [comment]
Bob got cancelled. 

![borek-stavitel-standing-i3053](https://user-images.githubusercontent.com/23278552/147827943-c677335e-9992-474d-ba0d-04a8a24bd30b.jpg)

**Comment by GreatApo** [comment]
I can fix it!

**Comment by FrutyX** [comment]
Can you?!

![errors](https://user-images.githubusercontent.com/23278552/147828067-58fd4e66-1be2-4b0c-ab81-e116386d510e.PNG)

**Comment by FrutyX** [comment]
@GreatApo I am not 100% sure, but it seems, it may be appearing only in Turma fights.

**Comment by GreatApo** [comment]
> @GreatApo I am not 100% sure, but it seems, it may be appearing only in Turma fights.

I know what the problem is... The * in that player's name is a regular expression character and then I am using it in a match, thus I need to replace every special character (most are not allowed in names anyway).
Try now, just refresh the report page...

**Comment by FrutyX** [comment]
@GreatApo start counting rounds :D.

**Comment by GreatApo** [comment]
Fixed for sure.

**Comment by FrutyX** [comment]
![buuuug](https://user-images.githubusercontent.com/23278552/147830144-c2738dec-a354-42fc-83df-64772fc5b16a.PNG)

**Comment by FrutyX** [comment]
I am telling ya, this occurs only in Turma, it's fine in arena.

**Comment by GreatApo** [comment]
> I am telling ya, this occurs only in Turma, it's fine in arena.

Sorry, there was one more place that I needed to escape the string.

**Comment by FrutyX** [comment]
At last! It seems to be fixed, I'll keep my eye 👁️  on the error log, as usually.

**Comment by FrutyX** [comment]
Very random, but already logged twice.

`Uncaught TypeError: Cannot read properties of undefined (reading 'style')`
`https://s34-cz.gladiatus.gameforge.com/game/index.php?mod=reports&t=2&sh=`

`core/source/reports.js?4.3.5&built=1641135158985:227`
it's
`line[0].getElementsByTagName('th')[2].style.textAlign = "right";`

**Comment by FrutyX** [comment]
Didn't saw this error again, interesting.

Trivial.

**Comment by FrutyX** [comment]
Oh, what we have here, a fix that did not work.

Unfortunately I have no clue when and why this occurs.

Uncaught TypeError: Cannot read properties of undefined (reading 'dataset')
`https://s34-cz.gladiatus.gameforge.com/game/index.php?mod=guildStorage&subsub=0&sh=`

`core/source/guild.storage.js?4.3.6&built=1641231942319:180`
it's
`if (item && typeof(item[0].dataset.gcaFlag_doubleClickEvent) !== 'undefined') delete item[0].dataset.gcaFlag_doubleClickEvent;`

**Comment by FrutyX** [comment]
I love when I find the reason just when I write that I don't know the reason.

@GreatApo! This errors shows up, when you use the item directly on an item, from your storage, you get me right? 

Items like grindstone, shred of leather, protective gear and this kind of stuff, anything that can be moved directly on the item.

**Comment by FrutyX** [comment]
For the record.


https://user-images.githubusercontent.com/23278552/150957521-44da1a3c-46cc-4b56-ba33-506a9b58080d.mp4



I have also some good news, other than this, my browser did not logged any other errors so far, it's clear.

**Comment by FrutyX** [comment]
I found a new one.

This one is appearing when your underworld copy drops something, GCA fails to detect the enemy and throws an error.

![bug](https://user-images.githubusercontent.com/23278552/175813022-c4e68334-b66e-4af3-b293-8e38f98c7649.PNG)

![bug2](https://user-images.githubusercontent.com/23278552/175813112-783d4325-e12e-4068-b70c-f8d6da656289.PNG)


`<div id="costumeAnimation_62b849e548476" class="avatar avatar_costume_part avatar_costume_animation" style="z-index: 0; background-image: url("9425/img/costumes/sets/male/3_complete.png"); background-position: -1848px 0px;"></div>`

**Comment by FrutyX** [comment]
It also seems it's only for secondary drops, primary was fine:

![bug3](https://user-images.githubusercontent.com/23278552/175813263-fc54027d-60b9-4314-8989-52ca1d4d897e.PNG)

No error here.

**Comment by FrutyX** [comment]
Since this issue is getting very long I won't be adding anything more here and next time I'll create a new issue.

Currently (I believe) there are 2 unresolved random console errors.

**Comment by GramThanos** [comment]
Its the addon trying to detect the materials drioped by the enemies. Here the enemy is yourself, thus it doesn't have an NPC number.

**Comment by FrutyX** [comment]
I said I will create a new issue but eh, scratch that, not needed tbh.

Category 16? I've checked the URL, the category was completely missing, it's not present on the page (it's the event titles category), it came back after vising again, weird behavior, perhaps you'll know more.

![cat](https://user-images.githubusercontent.com/23278552/177524456-2bc7ebb1-e838-4c26-baaf-498bccdf3873.PNG)

It could be just random, one of the ones that needs to be loaded perhaps, I wasn't researching the code so...
Still, better to let you know, I think I have met category 16 error in the past, but it's a long time ago and I am not sure anymore.

**Comment by FrutyX** [comment]
Another thing is... should it be at 0%, when I have one title acquired? I didn't even noticed until now that the percentages are a GCA feature.

![category](https://user-images.githubusercontent.com/23278552/177525446-94695165-473c-474a-8bee-2d29546d0ea3.PNG)

**Comment by FrutyX** [comment]
Tell me if you found something useful.

**Comment by FrutyX** [comment]
From now on, please continue in #368, #369 and #370.
